render-ws/src/main/scripts/docker-compose/02_docker_swarm.sh

```diff
@@ -0,0 +1,124 @@
+#!/bin/bash
+
+# This script is for running the render-ws containers in a docker swarm.
+
+set -e
+
+ABSOLUTE_SCRIPT=$(readlink -m "${0}")
+LOCAL_RENDER_WS_BASE_DIR=$(dirname "${ABSOLUTE_SCRIPT}")
+
+# ---------------------------------------------
+# edit these variables for the current server
+
+export JETTY_RUN_AS_USER_AND_GROUP_IDS="999:999"
+
+# ---------------------------------------------
+# everything below here should be left as is
+
+SWARM_STACK_NAME="render-swarm"
+BASENAME=$(basename "${0}")
+USAGE="USAGE:
+  ${BASENAME} up <render-ws-image>
+  ${BASENAME} up <render-ws-image repository> <render-ws-image tag>
+
+  Starts the specified image (e.g. janelia-render-ws:latest-ws) in a docker swarm.
+
+  The 3-parameter <repository> <tag> variant allows for easy cut+paste from the output of docker images.
+
+  The superfluous 'up' argument is required to make it clear that this script only
+  starts up containers (or shows this usage message).
+  Containers should only be brought up after the swarm has been initialized and hosts have joined.
+
+  Other useful docker commands (not handled by this script):
+
+  - Initialize the swarm with the current host as the manager (also prints join commands for other hosts):
+      # see https://docs.docker.com/engine/swarm/swarm-mode/
+      docker swarm init
+
+  - Force the current host to leave the swarm (forcing the last master to leave will destroy the swarm):
+      # see https://docs.docker.com/engine/reference/commandline/swarm_leave/
+      docker swarm leave --force
+
+  - Get the join token for a worker to join the swarm:
+      # see https://docs.docker.com/engine/reference/commandline/swarm_join-token/
+      docker swarm join-token worker
+      
+  - Stop render-ws containers on each host in the swarm:
+      # see https://docs.docker.com/engine/reference/commandline/stack_rm/
+      docker stack down ${SWARM_STACK_NAME}
+
+  - List the nodes in the swarm:
+      # see https://docs.docker.com/engine/reference/commandline/node_ls/
+      docker node ls
+
+  - List the tasks running on the current node in the swarm:
+      # see https://docs.docker.com/engine/reference/commandline/node_ps/
+      docker node ps
+
+  - List the services running in the swarm:
+      # see https://docs.docker.com/engine/reference/commandline/service_ls/
+      docker service ls
+"
+
+if (( $# < 2 )) || [ "$1" != "up" ]; then
+  echo "${USAGE}"
+  exit 1
+fi
+
+if (( $# == 2 )); then
+  export RENDER_WS_IMAGE="${2}"
+else
+  export RENDER_WS_IMAGE="${2}:${3}"
+fi
+
+if [ "${JETTY_RUN_AS_USER_AND_GROUP_IDS}" == "999:999" ]; then
```

Comment:
It's not clear to me where `JETTY_RUN_AS_USER_AND_GROUP_IDS` would be changed from the default value to prevent this error from being thrown.

Reply:
The script is intended to by copied and (minimally) edited for each deployment environment.  This variable should be changed on line 13.  I'll improve the error message.